### ISSUE TITLE
Inquiry about the F1 and IOU result of AF task using spatial models

### ISSUE THREAD

**Issue by zengyugong** [zhaoyutim/TS-SatFire#7]
Hi Dr. Zhao,

Thanks for your great work!

I encountered some issues when I tried to reproduce the spatial models in your paper. 
1. I generated AF dataset (test/val/test) using dataset_gen_afba.py. 
2. Then, I trained baseline models provided in the run_spatial_model.py. (command like: python run_spatial_model.py -m unet -mode af -b 16 -r 1 -lr 0.0003 -ts 1 -it 1 -nc 8 -nh 4)
The problem is that the val_loss and _train_loss always go to negative values, and mdice always goes to >1. In the end, the F1 score and mIOU in the test set are always <0.5, which is much worse than the models' performance in the paper.

So, I think some modifications may be added before running the run_spatial_model.py script.

**Comment by zhaoyutim** [comment]
Hi Yugong,

Thank you for the inquiry, it sounds strange to me even with UNet and spatial model this happened. I suggest you to debug a bit when loading the image in the data loader. And if you find anything, please make a screenshot and help me to identify the issue.

Kind regards,
Yu Zhao

**Comment by zhaoyutim** [comment]
I suspect there might be some misalignment with the label indexes, can you first visualize the label and input image?

**Comment by zengyugong** [comment]
Hi Dr. Zhao,

Many thanks for your response.

For debugging, I tried to print some values of the training and validation sets in the terminal. 
1. The input image data shape was [batch_size, 8, 256, 256]
2. The label shape was [batch_size, 2(num_classes), 256, 256],  (here I used batch_size=16)
3. I printed unique values in the second dimension, which are these 2 num_classes. The result showed that values range from 0-380, **where the 1st class([ : ,0]) contains only [0,1] value, which I think should be the actual label for the AF task**. The 2nd class ([ : ,1]) contains values from 264 to 380, which I think is suitable for the BA task. The output in the terminal was shown as:
Training set labels:

<img width="767" height="239" alt="Image" src="https://github.com/user-attachments/assets/fcf74edc-d219-4599-9770-d646355a6f5d" />

Also, similar to Val set labels:

<img width="779" height="243" alt="Image" src="https://github.com/user-attachments/assets/260124c6-134a-47e5-87cf-a07d12c1f3bc" />

In order to keep consistent with your original code, I still used num_classes=2 to train the Unet. **Then, the confusing issue came out here, even in the 1st epoch, the train loss and val loss went to negative**.

<img width="1893" height="86" alt="Image" src="https://github.com/user-attachments/assets/12fc68fc-700c-4cc1-a959-b46c503e4749" />

<img width="649" height="14" alt="Image" src="https://github.com/user-attachments/assets/3e671a59-07ce-4c35-9549-f6c15fdd56ee" />

Best regards,
Yugong Zeng

**Comment by zhaoyutim** [comment]
Hi Yugong,

To me the label values are strange, the value above 260 seems from VIIRS band I4. The labels should be binary. There might be some bugs during the selection of the spectral bands. I suggest you to visualize one data sample using QGIS or plot it out. Then using the correct index. And you are right, I think during the creation of the dataset, you should select the channel that belongs to the label and [:, 0] should be non-burned area and [:, 1] should be burned area, for example. 

Kind regards,
Yu Zhao

**Comment by zengyugong** [comment]
Hi Dr. Zhao,

Thanks for your valuable suggestion. 

I picked up some random samples (index 51, 52) from the val set and plotted, with the 8 input channels in image data (Frame 0-7) and the corresponding channels in label data (Label 0-2).

<img width="1605" height="590" alt="Image" src="https://github.com/user-attachments/assets/1a2db3b2-b12a-4d4b-b5c7-3b2751f597f5" />

<img width="1607" height="583" alt="Image" src="https://github.com/user-attachments/assets/ea972922-a627-4712-abf9-b5bd1bf06411" />

I think label channel 0 is all 0 (background), channel 1 is [0,1] masked, but channel 2 contains some values over 260.

**Comment by zhaoyutim** [comment]
Hi Yugong,

Thank you for your plots, I think in this case the channel 1 should be the burned area generated from accumulation of active fires and channel 2 should be the active fire if you set channel 2 > 0 to be the binary label.
Hope this help, I will later update the code but I think you can first try to generate the label using these two channels.

Kind regards,
Yu Zhao

**Comment by zhaoyutim** [comment]
The way to update the label selection is on Line 80 in data_generator_torch.py, in this case the ba should be set to 1 and af to be 2 for label_sel variable, it can be set in Line 95 in run_spatial_models.py. Also, I think it would be good to change the Line 79-80 in data_generator_torch to:

 y_dataset[0, :, :, :] = label_dataset[1, :, :, :] > 0
 y_dataset[1, :, :, :] = label_dataset[1, :, :, :] != 0

this corresponding to set the first channel to be burned area and the second channel to be non-burned area!

Hope it helps.

Kind regards,
Yu Zhao

**Comment by zengyugong** [comment]
Hi Dr. Zhao,

Many thanks for your reply.

I will follow your instructions. Any updates I'll post here.

Best regards,
Yugong Zeng

**Comment by zengyugong** [comment]
Hi Dr. Zhao,

I still have some questions.

1. At line 126-128 in run_spatial_model.py, the default setting in the train and val stage is 'include_background=True'. Also, the default num_classes = 2, I think it includes the background and the AF pixels. I tried with this, and the mIOU got really high (> 0.7 in val set). 
2. However, in the test stage, from line 288-299, we only use label[1, : , : ] to evaluate, which does not include the background. And I got a really low mIOU and mDICE (<0.4).
3. I turned the setting to 'include_background=False' and tried again, the mIOU got lower (about 0.5 in val set). But the mIOU was still <0.4 in testing.

<img width="691" height="555" alt="Image" src="https://github.com/user-attachments/assets/1b112e7c-8f16-4c5b-a838-46c4fae2d407" />

So I think there are maybe some misalignments in the test and train stage. The command I used is 'run_spatial_model.py -m attunet -mode af -b 16 -r 1 -lr 0.0003 -ts 1 -it 1 -nc 8 -nh 8 -ed 128'

Best regards,
Yugong Zeng

**Comment by zhaoyutim** [comment]
Hi Yugong,

Can you do the same as what you presented above to plot the testset to check whether the label channel you selected are the same as the channel in the train and validation? Since the test dataset was collected way before train and validation dataset, there might be misalignment in its order of bands. Sorry about the hassle.

Kind regards,
Yu Zhao

**Comment by zengyugong** [comment]
Hi Dr. Zhao,

Thanks for your reply.

I plotted the blue_ridge_fire data (af_blue_ridge_fire_img_seqtoseql_1i_1.npy, af_blue_ridge_fire_label_seqtoseql_1i_1.npy). 
It contains 10 days' fire images. Here I used the first 3. 

<img width="1610" height="639" alt="Image" src="https://github.com/user-attachments/assets/778603ff-0771-4f86-900d-b9920e90aecb" />

<img width="1586" height="604" alt="Image" src="https://github.com/user-attachments/assets/00177a9b-0004-42f5-a9f2-f18591ca09dc" />

<img width="1589" height="609" alt="Image" src="https://github.com/user-attachments/assets/1e30edde-c68f-47b7-9fd1-4d68195b256d" />

From the images above, I think the generated labels in different channels are all the same.
It's really different from the train and val set. 

Best regards,
Yugong Zeng

**Comment by zhaoyutim** [comment]
Hi Yugong,

Thank you for the info, I think either one should be fine in this case for testset, I intended to make it the same to keep the band number to be consistent. So in this case the label should be using either of them labels[0,...]>0 for af label, and labels[0,...]==0 for non-af labels.

I will also update the data processor sometime off my work. Sorry I might be only able to answer questions for now as I no longer work on this topic.

Kind regards,
Yu Zhao
